Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -103,28 +97,39 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
         return;
     }
 
-    if (!m_cookieJar) {
-        promise->reject(SecurityError);
-        return;
-    }
-
-    auto& document = *downcast<Document>(context);
-    auto completionHandler = [promise = WTFMove(promise)] (std::optional<Vector<Cookie>>&& cookies) {
-        if (!cookies) {
-            promise->reject(TypeError);
+    auto getCookies = [options = WTFMove(options), promise = WTFMove(promise)](ScriptExecutionContext& context) {
```

Comment:
We most probably do not want to capture the promise here, there is a risk it gets unrefed in main thread.
You can probably use a Vector or a HashMap in CookieStore to store it and retrieve it as part of the completion handler.
Or we should ensure the completion handler to always be destroyed in the creation thread, see how WorkerCacheStorageConnection does it.
One possibility would be for m_cookieJar to be the one handling whether it is used by a service worker or a document.

Hoping to main thread with options probably requires isolate copying options as well.
With auto getCookies = lambda, we had issues in the past where getCookies would be copy constructed without us noticing it, resulting in thread safety issues. It might be best to type it as a Function from the start.

Reply:
I have fixed the promise issue by using a HashMap in CookieStore!